url80587fc7d5_dennervidal_CloudSim_DVFS_tgz-pJ8-Examples_Dvfs_example_simpleJ8/src/org/cloudbus/cloudsim/network/datacenter/Switch.java

```diff
@@ -0,0 +1,386 @@
+				List<NetworkPacket> pktlist = packetTohost.get(hostid);
+				if (pktlist == null) {
+					pktlist = new ArrayList<NetworkPacket>();
+					packetTohost.put(hostid, pktlist);
+				}
+				pktlist.add(hspkt);
+				return;
+
+			}
+			// packet is to be sent to upper switch
+			// ASSUMPTION EACH EDGE is Connected to one aggregate level switch
+
+			Switch sw = uplinkswitches.get(0);
+			List<NetworkPacket> pktlist = uplinkswitchpktlist.get(sw.getId());
+			if (pktlist == null) {
+				pktlist = new ArrayList<NetworkPacket>();
+				uplinkswitchpktlist.put(sw.getId(), pktlist);
+			}
+			pktlist.add(hspkt);
+			return;
+		}
+		if (level == NetworkConstants.Agg_LEVEL) {
+			// packet is coming from edge level router so need to be sent to
+			// either root or another edge level swich
+			// find the id for edgelevel switch
+			int switchid = dc.VmToSwitchid.get(recvVMid);
+			boolean flagtoswtich = false;
+			for (Switch sw : downlinkswitches) {
+				if (switchid == sw.getId()) {
+					flagtoswtich = true;
+				}
+			}
+			if (flagtoswtich) {
+				List<NetworkPacket> pktlist = downlinkswitchpktlist.get(switchid);
+				if (pktlist == null) {
+					pktlist = new ArrayList<NetworkPacket>();
+					downlinkswitchpktlist.put(switchid, pktlist);
+				}
+				pktlist.add(hspkt);
+			} else// send to up
+			{
+				Switch sw = uplinkswitches.get(0);
+				List<NetworkPacket> pktlist = uplinkswitchpktlist.get(sw.getId());
+				if (pktlist == null) {
+					pktlist = new ArrayList<NetworkPacket>();
+					uplinkswitchpktlist.put(sw.getId(), pktlist);
+				}
+				pktlist.add(hspkt);
+			}
+		}
+		if (level == NetworkConstants.ROOT_LEVEL) {
+			// get id of edge router
+			int edgeswitchid = dc.VmToSwitchid.get(recvVMid);
+			// search which aggregate switch has it
+			int aggSwtichid = -1;
+			;
+			for (Switch sw : downlinkswitches) {
+				for (Switch edge : sw.downlinkswitches) {
+					if (edge.getId() == edgeswitchid) {
+						aggSwtichid = sw.getId();
+						break;
+					}
+				}
+			}
+			if (aggSwtichid < 0) {
+				System.out.println(" No destination for this packet");
+			} else {
+				List<NetworkPacket> pktlist = downlinkswitchpktlist.get(aggSwtichid);
+				if (pktlist == null) {
+					pktlist = new ArrayList<NetworkPacket>();
+					downlinkswitchpktlist.put(aggSwtichid, pktlist);
+				}
+				pktlist.add(hspkt);
+			}
+		}
+	}
+
+	private void registerHost(SimEvent ev) {
+		NetworkHost hs = (NetworkHost) ev.getData();
+		hostlist.put(hs.getId(), (NetworkHost) ev.getData());
+	}
+
+	protected void processpacket(SimEvent ev) {
+		// send packet to itself with switching delay (discarding other)
+		CloudSim.cancelAll(getId(), new PredicateType(CloudSimTags.Network_Event_UP));
+		schedule(getId(), switching_delay, CloudSimTags.Network_Event_UP);
+		pktlist.add((NetworkPacket) ev.getData());
+
+		// add the packet in the list
+
+	}
+
+	private void processOtherEvent(SimEvent ev) {
+
+	}
+
+	protected void processpacketforward(SimEvent ev) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 13. By comparison, 98% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 283-299, lines 300-316, lines 317-332 into separate methods.